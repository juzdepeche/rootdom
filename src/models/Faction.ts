import images from '../../assets/images';

export class Faction {
	public name: string;
	public value: number;
	public color: string;
	public image;

	constructor(name: string, value: number, color: string, image?) {
		this.name = name;
		this.value = value;
		this.color = color;
		this.image = image;
	}
}

export const Factions = [
	new Faction('Marquise de Cat', 10, '#FD5200', images.cat),
	new Faction('Underground Duchy', 8, '#D7A872', images.taupe),
	new Faction('Eyrie Dynasties', 7, '#4774B7', images.bird),
	new Faction('Vagabond', 5, '#FCFEFD', images.vagabond),
	new Faction('Riverfolk Company', 5, '#61DDFB', images.otter),
	new Faction('Woodland Alliance', 3, '#59BA53', images.mouse),
	new Faction('Corvid Consipiracy', 3, '#290D7E', images.raven),
	new Faction('Lizard Cult', 2, '#f1f435', images.lizard)
];

export const GetViableGameSum = (playerCount:number) => {
	switch (playerCount) {
		case 2:
			return 17;
		case 3:
			return 18;
		case 4:
			return 21;
		case 5:
			return 25;
		case 6:
			return 28;
		default:
			return 0;
	}
}

export const GetBigFactionsSum = (factions:Faction[], playerCount:number) => {
	let sum = 0;
	factions.sort((a, b) => a.value > b.value ? -1 : 1)
	for (let i = 0; i < playerCount; i++) {
		sum += factions[i].value;
	}
	return sum;
}