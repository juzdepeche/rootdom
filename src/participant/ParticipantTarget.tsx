import React, {
	useEffect,
	useState,
	forwardRef,
	useImperativeHandle
} from 'react';
import { View, StyleSheet, Animated, Image, Text } from 'react-native';
import { Faction } from '../../App';

const colors = [
	'red',
	'blue',
	'yellow',
	'green',
	'orange',
	'cyan',
	'plum',
	'gray',
	'purple'
];
const TOUCH_SIZE = 100;
const TIME_NEEDED = 1500;
const SCALE_ANIMATION_DURATION = 1500;
const SHRINK_ANIMATION_DURATION = 100;
const SHRINK_ANIMATION_DURATION_RANGE = 250;

export type ParticipantTargetProps = {
	item;
	index;
	onReadyCallback;
	dispatched: boolean;
	faction: Faction;
};

export const ParticipantTarget = forwardRef(
	(
		{
			item,
			index,
			onReadyCallback,
			dispatched,
			faction
		}: ParticipantTargetProps,
		ref
	) => {
		const scale = 1 + (item.force || 0) * 2;
		const startOpacity = 0.6;

		const [pulsePercantageValue] = useState(new Animated.Value(scale));
		const [opacityPercentageValue] = useState(
			new Animated.Value(startOpacity)
		);
		const [animation, setAnimation] = useState(
			{} as Animated.CompositeAnimation
		);

		const startTimer = () => {
			return setTimeout(() => {
				onReadyCallback(index);
			}, TIME_NEEDED);
		};

		const startScaleAndOpacityAnimation = () => {
			Animated.timing(opacityPercentageValue, {
				toValue: 1,
				duration: TIME_NEEDED
			}).start();

			const pulseAnimation = Animated.timing(pulsePercantageValue, {
				toValue: 3,
				duration: SCALE_ANIMATION_DURATION
			});

			setAnimation(pulseAnimation);

			Animated.loop(pulseAnimation).start();
		};

		useEffect(() => {
			const handler = startTimer();
			startScaleAndOpacityAnimation();

			//when componenet is destroyed
			return () => {
				clearTimeout(handler);
			};
		}, []);

		useImperativeHandle(ref, () => ({
			shrink(destroyCallback: (index) => void) {
				animation.stop();
				Animated.spring(pulsePercantageValue, {
					toValue: 0,
					speed: 20
				}).start(() => destroyCallback(index));
			}
		}));

		const pulsePercentage = pulsePercantageValue.interpolate({
			inputRange: [0, 1, 1.9, 3],
			outputRange: [0, 1, 1.1, 1]
		});

		const opacityPercentage = opacityPercentageValue.interpolate({
			inputRange: [startOpacity, 1],
			outputRange: [0.6, 1]
		});

		return (
			<View>
				{/* {!dispatched ? null : (
					<Text
						style={[
							styles.text,
							{
								top: item.pageY - TOUCH_SIZE,
								left: item.pageX - TOUCH_SIZE / 2
							}
						]}
					>
						{faction.name}
					</Text>
				)} */}
				<Animated.View
					style={[
						styles.touch,
						{
							transform: [
								{ translateX: -TOUCH_SIZE / 2 },
								{ translateY: -TOUCH_SIZE / 2 },
								{ scale: pulsePercentage }
							],
							opacity: opacityPercentage,
							backgroundColor: faction?.color
								? faction?.color
								: colors[index % colors.length],
							top: item.pageY,
							left: item.pageX
						}
					]}
				>
					{faction?.image ? (
						<Image source={faction?.image} style={styles.image} />
					) : null}
				</Animated.View>
			</View>
		);
	}
);

const styles = StyleSheet.create({
	touch: {
		position: 'absolute',
		aspectRatio: 1,
		width: TOUCH_SIZE,
		borderRadius: TOUCH_SIZE / 2
	},
	text: {
		position: 'absolute',
		color: 'white',
		zIndex: 10,
		fontSize: 20,
		alignItems: 'center',
		backgroundColor: 'blue'
	},
	image: {
		width: TOUCH_SIZE,
		height: TOUCH_SIZE
	}
});
