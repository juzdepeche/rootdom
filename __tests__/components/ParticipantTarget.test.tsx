import React from 'react';
import renderer from 'react-test-renderer';
import { Factions } from '../../src/models/Faction';
import ParticipantTarget from '../../src/components/ParticipantTarget';

jest.useFakeTimers();

describe('<ParticipantTarget />', () => {
	it('renders correctly', () => {
		const key = 1;
		const ref = { current: {} } as any;
		const touch = { pageX: 100, pageY: 200 };
		const onReadyCallback = () => true;
		const dispatched = false;
		const faction = Factions[0];
		spyOn(React, 'createRef').and.returnValue(ref);
		const participantTarget = renderer
			.create(
				<ParticipantTarget
					key={key}
					ref={ref}
					item={touch}
					index={key}
					onReadyCallback={onReadyCallback}
					dispatched={dispatched}
					faction={faction}
				></ParticipantTarget>
			)
			.toJSON();
		expect(participantTarget).toMatchSnapshot();
	});
});
