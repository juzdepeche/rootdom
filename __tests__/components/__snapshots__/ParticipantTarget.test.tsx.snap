// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParticipantTarget /> renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "aspectRatio": 1,
        "backgroundColor": "#FD5200",
        "borderRadius": 50,
        "left": 100,
        "opacity": 0.6,
        "position": "absolute",
        "top": 200,
        "transform": Array [
          Object {
            "translateX": -50,
          },
          Object {
            "translateY": -50,
          },
          Object {
            "scale": 1,
          },
        ],
        "width": 100,
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </View>
</View>
`;
